addons/procurement/procurement.py

```diff
@@ -60,177 +52,143 @@ class procurement_rule(osv.osv):
+    rule_id = fields.Many2one('procurement.rule', string='Rule', track_visibility='onchange', help="Chosen rule for the procurement resolution. Usually chosen by the system but can be manually set by the procurement manager to force an unusual behavior.")
+    product_id = fields.Many2one('product.product', string='Product', required=True, states={'confirmed': [('readonly', False)]}, readonly=True)
+    product_qty = fields.Float(string='Quantity', digits_compute=dp.get_precision('Product Unit of Measure'), required=True, states={'confirmed': [('readonly', False)]}, readonly=True)
+    product_uom = fields.Many2one('product.uom', string='Product Unit of Measure', required=True, states={'confirmed': [('readonly', False)]}, readonly=True)
+    state = fields.Selection([('cancel', 'Cancelled'), ('confirmed', 'Confirmed'), ('exception', 'Exception'), ('running', 'Running'), ('done', 'Done')], string='Status', required=True, track_visibility='onchange', copy=False, default='confirmed')
+
+    @api.model
+    def _needaction_domain_get(self):
+        return [('state', '=', 'exception')]
+
+    @api.multi
+    def unlink(self):
+        unlink_ids = self.filtered(lambda procurements: procurements.state == 'cancel')
+        for s in (self-unlink_ids):
+            if s.state != 'cancel':
+                raise UserError(_('Cannot delete Procurement Order(s) which are in %s state.') % s.state)
+        return super(ProcurementOrder, unlink_ids).unlink()
+
+    @api.multi
+    def do_view_procurements(self):
         '''
         This function returns an action that display existing procurement orders
         of same procurement group of given ids.
         '''
-        act_obj = self.pool.get('ir.actions.act_window')
-        action_id = self.pool.get('ir.model.data').xmlid_to_res_id(cr, uid, 'procurement.do_view_procurements', raise_if_not_found=True)
-        result = act_obj.read(cr, uid, [action_id], context=context)[0]
-        group_ids = set([proc.group_id.id for proc in self.browse(cr, uid, ids, context=context) if proc.group_id])
+        result = self.env['ir.actions.act_window'].for_xml_id('procurement', 'do_view_procurements')
+        group_ids = set([proc.group_id.id for proc in self if proc.group_id])
         result['domain'] = "[('group_id','in',[" + ','.join(map(str, list(group_ids))) + "])]"
         return result
 
-    def onchange_product_id(self, cr, uid, ids, product_id, context=None):
+    @api.onchange('product_id')
+    def onchange_product_id(self):
         """ Finds UoM of changed product.
-        @param product_id: Changed id of product.
-        @return: Dictionary of values.
         """
-        if product_id:
-            w = self.pool.get('product.product').browse(cr, uid, product_id, context=context)
-            v = {
-                'product_uom': w.uom_id.id,
-            }
-            return {'value': v}
-        return {}
+        if self.product_id:
+            self.product_uom = self.product_id.uom_id.id
```

Comment:
No need to check the condition and also better to assign recordset(s) instead of id(s).